### DISCUSSION TITLE
Middleware for forwarded messages

### DISCUSSION THREAD

**Discussion by RmzzzCode** [aiogram/aiogram#1402] [Q&A]
Hi guys! I had difficulty writing middleware for forwarded messages. The bottom line is that I am sending several forwarded messages to the bot and it should read them as one message. I searched the Internet for information, but all I found was middleware for media_group, I'll attach the code below. If anyone knows or has come across, help, I will be grateful!
```python
import asyncio
from typing import Union
from aiogram import types
from aiogram.dispatcher.handler import CancelHandler
from aiogram.dispatcher.middlewares import BaseMiddleware


class AlbumMiddleware(BaseMiddleware):

    album_data: dict = {}

    def __init__(self, latency: Union[int, float] = 0.01):
        
        self.latency = latency
        super().__init__()

    async def on_process_message(self, message: types.Message, data: dict):
        if not message.media_group_id:
            return

        try:
            self.album_data[message.media_group_id].append(message)
            raise CancelHandler() 
        except KeyError:
            self.album_data[message.media_group_id] = [message]
            await asyncio.sleep(self.latency)

            message.conf["is_last"] = True
            data["album"] = self.album_data[message.media_group_id]

    async def on_post_process_message(self, message: types.Message, result: dict, data: dict):
        
        if message.media_group_id and message.conf.get("is_last"):
            del self.album_data[message.media_group_id]
```

**Comment by franklin-lol** [comment]
Hello, sir!
Forwarded messages don't have a shared media_group_id, so you can't group them the same way. The key is forward_origin — all forwarded messages from the same original chat arrive as separate updates, but you can group them by (user_id, forward_origin.chat.id) with a short collection window:
pythonimport asyncio
from collections import defaultdict
from typing import Any, Awaitable, Callable
from aiogram import BaseMiddleware
from aiogram.types import Message

class ForwardGroupMiddleware(BaseMiddleware):
    def __init__(self, latency: float = 0.3):
        self.latency = latency
        self._groups: dict[tuple, list[Message]] = defaultdict(list)
        self._tasks: dict[tuple, asyncio.Task] = {}

    async def __call__(
        self,
        handler: Callable[[Message, dict[str, Any]], Awaitable[Any]],
        event: Message,
        data: dict[str, Any],
    ) -> Any:
        # only intercept forwarded messages
        if not event.forward_origin:
            return await handler(event, data)

        origin = event.forward_origin
        origin_id = getattr(getattr(origin, "chat", None), "id", None) or \
                    getattr(getattr(origin, "sender_user", None), "id", None)
        key = (event.from_user.id, origin_id)

        self._groups[key].append(event)

        # cancel existing timer, restart it
        if key in self._tasks:
            self._tasks[key].cancel()

        async def flush():
            await asyncio.sleep(self.latency)
            messages = self._groups.pop(key, [])
            self._tasks.pop(key, None)
            if messages:
                data["forwarded_group"] = messages
                await handler(messages[0], data)

        self._tasks[key] = asyncio.create_task(flush())
Register:
pythondp.message.middleware(ForwardGroupMiddleware(latency=0.4))
In your handler:
python@dp.message()
async def handle(message: Message, forwarded_group: list[Message] = None):
    if forwarded_group:
        print(f"Got {len(forwarded_group)} forwarded messages")
The latency window (0.3–0.5s) gives Telegram time to deliver all messages from the batch before processing.
